app/controllers/collections_controller.rb

```diff
@@ -167,25 +167,78 @@ def add_items_to_collection
     begin
       request_params = cleanse_params(params)
       collection = validate_collection(request_params[:id], request_params[:api_key])
-      corpus_dir = api_corpus_dir(collection.name)
-      validate_add_items_request(collection, corpus_dir, request_params[:items], request_params[:file])
+      validate_add_items_request(collection, collection.corpus_dir, request_params[:items], request_params[:file])
       request_params[:items].each do |item_json|
         update_ids_in_jsonld(item_json["metadata"], collection)
       end
-      uploaded_files = process_uploaded_files(corpus_dir, collection.name, request_params[:file])
-      processed_items = process_items(collection.name, corpus_dir, request_params, uploaded_files)
-      create_collection_manifest(corpus_dir)
-      @success_message = ingest_items(corpus_dir, processed_items[:successes]) # Respond with a list of items added (via item ingest)
+      uploaded_files = process_uploaded_files(collection.corpus_dir, collection.name, request_params[:file])
+      processed_items = process_items(collection.name, collection.corpus_dir, request_params, uploaded_files)
+      @success_message = add_item_core(collection, processed_items[:successes])
     rescue ResponseError => e
       respond_with_error(e.message, e.response_code)
       return # Only respond with one error at a time
     end
   end
 
+  #
+  # Core functionality common to add item ingest (via api and web app)
+  # Returns a list of item identifiers corresponding to the items ingested
+  #
+  def add_item_core (collection, item_id_and_file_hash)
+    rdf_files = []
+    item_id_and_file_hash.each do |item|
+      rdf_files.push(item[:rdf_file])
+    end
+    update_collection_manifest(collection.corpus_dir, rdf_files)
+    ingest_items(collection.corpus_dir, item_id_and_file_hash)
+  end
+
+  def web_add_item
```

Comment:
It all looks good, but this method is a little big still. You could decompose this a little be more. You've grouped some of the statements logically and commented on them, e.g.
```
 # Check required fields
required_fields = {:item_name => 'item name', :item_title => 'item title'}
  required_fields.each do |key, value|
  raise ResponseError.new(400), "Required field '#{value}' is missing" if params[key].blank?
end
```
You could make this a method, and with a good method name you won't need the comment because the reader would have a good assumption about what it does:

```
check_required_fields(params)
```

You could probably do the same with these comments:

```
# Check item name is unique within the collection
# Construct item Json-ld
```